### PR TITLE
[Off-chain, Testing] test: protocol pkg (part 1)

### PR DESCRIPTION
**Pull request by bryanchriswhite**
### Human Summary

- Refactors `BytesDifficultyGreaterThan()` into `CountDifficultyBits()` (and `MustCountDifficultyBits()`) with bit-level resolution (as opposed to bytes) :sunglasses: 
- Adds test coverage around success and error paths of `CountDifficultyBits()`.

### AI Summary

reviewpad:summary

## Issue

- #126

Unit test coverage is missing for large portions of `pkg/relayer/...`.

## Type of change

Select one or more:

- [ ] New feature, functionality or library
- [ ] Bug fix
- [x] Code health or cleanup
- [ ] Documentation
- [ ] Other (specify)

## Testing

- [ ] **Run all unit tests**: `make go_develop_and_test`
- [ ] **Verify Localnet manually**: See the instructions [here](TODO: add link to instructions)

## Sanity Checklist

- [ ] I have tested my changes using the available tooling
- [x] I have performed a self-review of my own code
- [ ] I have commented my code, updated documentation and left TODOs throughout the codebase

### CODE REVIEW

File: pkg/relayer/protocol/difficulty.go [hunk 1]
```diff
@@ -1,17 +1,40 @@
...
+	return diff
+}
+
+// CountDifficultyBits returns the number of leading zero bits in the given byte
+// slice. It returns an error if the byte slice is all zero bits.
+func CountDifficultyBits(bz []byte) (int, error) {
+	bzLen := len(bz)
+
+	var zeroBits int
+	for i, b := range bz {
```
Comment (Olshansk) [inline pkg/relayer/protocol/difficulty.go:27]:
```suggestion
	for byteIdx, byteValue := range bz {
```

File: pkg/relayer/protocol/difficulty.go [hunk 2]
```diff
@@ -1,17 +1,40 @@
...
+		if b != 0 {
+			zeroBits = bits.LeadingZeros8(b)
+			if zeroBits == 8 {
+				// we already checked that b != 0.
+				return 0, ErrDifficulty.Wrap("impossible code path")
+			}
+
+			return (i)*8 + zeroBits, nil
```
Comment (Olshansk) [inline pkg/relayer/protocol/difficulty.go:35]:
```suggestion
			// We have byteIdx bytes that are all 0s and one byte that is partially is with exactly zeroBits 
			return (byteIdx)*8 + zeroBits, nil
```

File: pkg/relayer/protocol/difficulty_test.go [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+			difficulty: 10,
+		},
+		{
+			bz:         []byte{0, 0, 255, 255},
+			difficulty: 16,
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(fmt.Sprintf("difficulty_%d", tt.difficulty), func(t *testing.T) {
```
Comment (Olshansk) [inline pkg/relayer/protocol/difficulty_test.go:44]:
```suggestion
		t.Run(fmt.Sprintf("difficulty_%d_zero_bits", tt.difficulty), func(t *testing.T) {
```

File: pkg/relayer/miner/miner.go [hunk 4]
```diff
@@ -102,7 +102,7 @@ func (mnr *miner) mapMineRelay(
 	relayHash := mnr.hash(relayBz)
 
 	// The relay IS NOT volume / reward applicable
-	if !protocol.BytesDifficultyGreaterThan(relayHash, defaultRelayDifficulty) {
+	if protocol.MustCountDifficultyBits(relayHash) >= defaultRelayDifficultyBits {
```
Comment (Olshansk) [inline pkg/relayer/miner/miner.go:105]: Should this be `<` given that this is the `NOT volume applicable` codepath?

**Review by Olshansk — CHANGES_REQUESTED** [review]
REALLY like this new approach. Left a couple small nits but not approving because I think our equivalence check may be wrong for the volume applicability codepath.

File: pkg/relayer/miner/miner.go (hunk 4 above)
Comment (bryanchriswhite) [reply pkg/relayer/miner/miner.go:105]: Nice catch, thanks! :facepalm: I guess it's been another long week. :sweat_smile:

Comment (bryanchriswhite) [reply pkg/relayer/miner/miner.go:105]: FWIW, the tests added in #191 would've caught this. :mechanical_arm:

File: e2e/tests/node.go [hunk 5]
```diff
@@ -117,7 +117,13 @@ func (p *pocketdBin) runPocketCmd(args ...string) (*commandResult, error) {
 func (p *pocketdBin) runCurlPostCmd(rpcUrl string, service string, data string, args ...string) (*commandResult, error) {
 	dataStr := fmt.Sprintf("%s", data)
 	urlStr := fmt.Sprintf("%s/%s", rpcUrl, service)
-	base := []string{"-v", "POST", "-H", "Content-Type: application/json", "--data", dataStr, urlStr}
+	base := []string{
```
Comment (bryanchriswhite) [inline e2e/tests/node.go:120]: I intended for this to go in with #178 but it got lost.

File: pkg/relayer/miner/miner.go (hunk 4 above)
Comment (Olshansk) [reply pkg/relayer/miner/miner.go]: I was going to just having a TODO to add tests for this regression but you're one step ahead!

**Review by Olshansk — APPROVED** [review]

**Pull request merged**